enhancements/installer/compact-plus-worker.md

```diff
@@ -0,0 +1,194 @@
+---
+title: Compact + Worker deployment as day0 operation
+authors:
+  - "@flaper87"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+api-approvers: # in case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers)
+  - TBD
+creation-date: 2022-03-08
+last-updated: 2022-03-08
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - TBD
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# Compact + Worker deployment as day0 operation
+
+## Summary
+
+This enhancement proposes to allow for deploying a 4-nodes cluster as a day 0
+operation, where the infrastructure workload can be deployed with HA enabled.
+The 4-nodes cluster would be composed by 3 schedulable master nodes and a
+single worker, which provide users with a highly available cluster.
+
+## Motivation
+
+There are environments where only 4 nodes are available where 4-nodes clusters
+are being deployed. We could serve OpenShift cluster administrators of these
+environments better by allowing them to deploy a 4-nodes cluster as a day0
+operation and make sure their infrastructure workload is highly available.
+
+Today, it is possible to deploy a 4-nodes cluster. The resulting cluster will
+be functional but the infrastructure workload will be run as a SingleReplica
+topology, which won't allow the various services to run as HA.
+
+To achieve the desired topology described in this enhancement,OpenShift cluster
+administrators can only deploy a 4-nodes clusters by first deploying a compact
+cluster (3 schedulable masters) and then adding a host as a day 2 operation.
+Another option is for them to This provides the same results as aimed in this
+enhancement but with a poor and slower user experience.
+
+Today, because of the current limitations in the installer, it takes
+administrators significantly longer to achieve the desired topology, when only
+4 nodes are available. Furthermore, by requiring more steps, we are making the
+process more prone to user errors.
+
```

Comment:
Does it make any difference what deployment approach or platform is being used?

Reply:
I think we're currently doing validation to prevent this configuration via IPI deployments on all platforms, right? So are we talking about removing that validation and doing something different? Or is this limited to the metal ZTP deployment scenarios?

Reply:
@dhellmann I actually don't think we're doing validation to prevent this (I just tried it for baremetal IPI and the installer accepted 1 worker replica)

So I think all we're talking about is extending the existing [scheduler manifest](https://github.com/openshift/installer/blob/master/pkg/asset/manifests/scheduler.go#L74) special-case so the controlplane nodes are schedulable when zero or 1 compute replicas are specified.

If we think this is specific to ZTP then it may be reasonable to just modify the installer-generated manifests (and this perhaps limits the test impact, e.g we probably don't want to test 3+1 for every platform in CI?)

Reply:
Ah, perhaps that validation was removed or changed when we added support for compact clusters. Or maybe I'm imagining it.

Reply:
> If we think this is specific to ZTP then it may be reasonable to just modify the installer-generated manifests (and this perhaps limits the test impact, e.g we probably don't want to test 3+1 for every platform in CI?)

I don't think this is ZTP specific. Let's ignore that this came from a ZTP environment and focus on the hardware we are dealing with. Imagine an enclosure that only has 4 blades in it. The option to add another blade is out of the question. This leaves the administrator with the option to either do the installation in 2 steps or have to go through the process of patching manifests. 

I understand the concern on the test impact of this change.